modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/set/IgniteMapSetOp.java

```diff
@@ -0,0 +1,107 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.query.calcite.rel.set;
+
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import com.google.common.collect.ImmutableList;
+import org.apache.calcite.plan.RelTraitSet;
+import org.apache.calcite.rel.core.CorrelationId;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.util.Pair;
+import org.apache.ignite.internal.processors.query.calcite.exec.exp.agg.AggregateType;
+import org.apache.ignite.internal.processors.query.calcite.exec.exp.agg.GroupKey;
+import org.apache.ignite.internal.processors.query.calcite.trait.CorrelationTrait;
+import org.apache.ignite.internal.processors.query.calcite.trait.IgniteDistributions;
+import org.apache.ignite.internal.processors.query.calcite.trait.RewindabilityTrait;
+import org.apache.ignite.internal.processors.query.calcite.trait.TraitUtils;
+import org.apache.ignite.internal.processors.query.calcite.type.IgniteTypeFactory;
+import org.apache.ignite.internal.processors.query.calcite.util.Commons;
+
+/**
+ * Physical node for MAP phase of set op (MINUS, INTERSECT).
+ */
+public interface IgniteMapSetOp extends IgniteSetOp {
+    /** {@inheritDoc} */
+    @Override public default List<Pair<RelTraitSet, List<RelTraitSet>>> deriveRewindability(
+        RelTraitSet nodeTraits,
+        List<RelTraitSet> inputTraits
+    ) {
+        boolean rewindable = inputTraits.stream()
+            .map(TraitUtils::rewindability)
+            .allMatch(RewindabilityTrait::rewindable);
+
+        if (rewindable)
+            return ImmutableList.of(Pair.of(nodeTraits.replace(RewindabilityTrait.REWINDABLE), inputTraits));
+
+        return ImmutableList.of(Pair.of(nodeTraits.replace(RewindabilityTrait.ONE_WAY),
+            Commons.transform(inputTraits, t -> t.replace(RewindabilityTrait.ONE_WAY))));
+    }
+
+    /** {@inheritDoc} */
+    @Override public default List<Pair<RelTraitSet, List<RelTraitSet>>> deriveDistribution(
+        RelTraitSet nodeTraits,
+        List<RelTraitSet> inputTraits
+    ) {
+        if (inputTraits.stream().allMatch(t -> TraitUtils.distribution(t).satisfies(IgniteDistributions.single())))
+            return ImmutableList.of(); // If all distributions are single or broadcast IgniteSingleMinus should be used.
+
+        if (inputTraits.stream().anyMatch(t -> TraitUtils.distribution(t) == IgniteDistributions.single()))
+            return ImmutableList.of(); // Mixing of single and random is prohibited.
```

Comment:
I think we discussed it here https://github.com/apache/ignite/pull/9009#discussion_r617440209 and come to the conclusion that only random distribution should be allowed for input nodes?

Reply:
> come to the conclusion that only random distribution should be allowed for input nodes

My suggestion doesn't violate this. You could require a RANDOM distribution for input nodes, and in case of SINGLE distributed child the Exchange with random distribution function should be created. In case of BROADCAST input we could require any HASH distribution, so the TrimExchange will be created.

Does it make any sense?

Reply:
So currently the test `testSetOpSingleAndRandom()` provides plane like this:
```
IgniteSingleMinus(all=[false])
  IgniteTableScan(table=[[PUBLIC, SINGLE_TBL1]])
  IgniteExchange(distribution=[single])
    IgniteTableScan(table=[[PUBLIC, RANDOM_TBL1]])
```

And follow plan should be possible too:
```
IgniteReduceMinus(all=[false])
  IgniteExchange(distribution=[single])
    IgniteMapMinus(all=[false])
      IgniteExchange(distribution=[random])
        IgniteTableScan(table=[[PUBLIC, SINGLE_TBL1]])
      IgniteTableScan(table=[[PUBLIC, RANDOM_TBL1]])
```

Reply:
I think this plan is not correct:
```IgniteReduceMinus(all=[false])
  IgniteExchange(distribution=[single])
    IgniteMapMinus(all=[false])
      IgniteExchange(distribution=[random])
        IgniteTableScan(table=[[PUBLIC, SINGLE_TBL1]])
      IgniteTableScan(table=[[PUBLIC, RANDOM_TBL1]])
```
This extra exchange with `distribution=[random]` is redundant, we don't need to send any data anywhere here, instead `single` distribution should satisfy `random` distribution and it makes possible plans like this:
 ```IgniteReduceMinus(all=[false])
  IgniteExchange(distribution=[single])
    IgniteMapMinus(all=[false])
      IgniteTableScan(table=[[PUBLIC, SINGLE_TBL1]])
      IgniteTableScan(table=[[PUBLIC, RANDOM_TBL1]])
```

Reply:
Fixed